### PR TITLE
formatters: Avoid changing name of the field

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
The current logic of the formatters changes the name of the original field (adding a "_raw" suffix) and creates a new field with the original name.

This behavior is not right as the name and type of the fields depends on the operators being run, hence, it could cause issues when parsing the output of Inspektor Gadget and enabling / disabling the formatters operator.

This commit fixes it by keeping the original name in all the cases. The new name is derived from an annotation on the metadata file.

## Testing 

Columns output prints the pretty version of the fields (sig and timestamp in this example)

```bash 
$ sudo -E ig run trace_signal:latest --verify-image=false
...
RUNTIME.CONTAINERNAME  PID          TPID         MNTNS_ID  UID          GID         RET COMM        SIG         TIMESTAMP
c3                     16770        85763        402653492 0            0           0   sh          SIGTERM     2024-06-17T09:40:55.4905
```

Raw fields are also available in the json output: 

```bash 
$ sudo -E ig run trace_signal:latest --verify-image=false -o json | jq
...
{
  "comm": "sh",
  "gid": 0,
  "k8s": {
    "containerName": "",
    "hostnetwork": false,
    "namespace": "",
    "node": "",
    "podName": ""
  },
  "mntns_id": 4026534922,
  "pid": 16770,
  "ret": 0,
  "runtime": {
    "containerId": "67df57854c8dd52b766560725f3daec0a2bc423f00875f58f1197ca3b34a68f5",
    "containerImageDigest": "",
    "containerImageName": "busybox",
    "containerName": "c3",
    "runtimeName": "docker"
  },
  "sig": "SIGTERM",
  "sig_raw": 15,
  "timestamp": "2024-06-17T09:41:44.057569398-05:00",
  "timestamp_raw": 1718635304057569300,
  "tpid": 85846,
  "uid": 0
}
```

### `_raw` suffix 

@alban suggested to use a `_raw` suffix and automatically trim it from the operators. It's good because avoids having to specify the annotation, but I don't like having this "magic" behavior, so this PR always use the annotation. Should I implement the raw suffic support too?

## TODO

- [ ] Future PR: Carry on annotations, attributes, description to created field

- Ref #3020
- Ref #2816 (Fixes part of it, however {l3,l4}endpoints are still TBD in another PR)

### CODE REVIEW

**Comment by alban** [comment]
@mauriciovasquezbernal I like the logic you have implemented :+1:

**Comment by flyth** [comment]
I think `field.targetname` is not specific enough for two reasons:

* there could be multiple targets for different converters/formatters/operators
* it's hard to find out what exactly reformatted the output by just looking at it and the metadata

I'd suggest something that is specific for the formatter/converter (in these cases) like `formatters.timestamp.target`, so you know instantly what operator and which functionality it affects.

> It's good because avoids having to specify the annotation, but I don't like having this "magic" behavior

I'm torn here: I also don't think too much magic is good, but this feels like something that is going to be used a LOT (timestamps especially) and I also think that we should try to avoid for the user to write a metadata file if not really necessary as it's the boring part of writing gadgets - in most cases the icing on the cake. And IMHO it also steepens the learning curve.

What about applying the "magic" only if additional circumstances are met, then? Like type AND name have to match - and maybe then issue a debug logging message saying "auto-generating 'timestamp' field based on name and type"? I know this adds a bit of logic in our code, but that way users can just copy/paste our gadget code and it would work ootb - and by seeing the message they'd know why it is there (should that question come up).

Difference in dev experience/learning curve for a first simple gadget:
* "oh nice, it shows the time human-readable automatically although it's an int"
* "Okay, that's an int, but I need the time in a readable format.
  * Let's debug: 'field.target annotation not found' - what is that? annotations? Have to read up on that. Oh, I need a metadata file. Oh, I need to describe my fields in yaml....".

But as said: I'm torn. Too much magic is not good. 🤷

**Comment by mauriciovasquezbernal** [comment]
I updated the logic to also use some "magic" when the field has the `raw_` suffix. Please let me know what you think.

File: pkg/utils/annotations/annotations.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		logger.Debugf("%s: using custom target field %q for field %q", logPrefix, outName, in.Name())
+		return outName, nil
+	}
+
+	if outName, ok := strings.CutSuffix(in.Name(), "_raw"); ok {
+		logger.Debugf("%s: using %q as target field for %q", logPrefix, outName, in.Name())
+		return outName, nil
+	}
+
+	return "", fmt.Errorf("%q annotation nor '_raw' prefix found", targetAnnotation)
```
Comment (flyth) [inline pkg/utils/annotations/annotations.go:43]:
```suggestion
	return "", fmt.Errorf("neither %q annotation nor '_raw' suffix found", targetAnnotation)
```

File: pkg/utils/annotations/annotations.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+)
+
+func GetTargetNameFromAnnotation(
+	logger logger.Logger,
+	logPrefix string,
+	in datasource.FieldAccessor,
+	targetAnnotation string,
```
Comment (flyth) [inline pkg/utils/annotations/annotations.go:29]: targetAnnotation could be replaced by logPrefix + `.target` (at least for the formatters pkg currently), but I guess you separated it to have the full annotation key (e.g. `timestampTargetAnnotation`) mentioned in code and thus searchable, right?

File: pkg/operators/ebpf/converters.go [hunk 3]
```diff
@@ -22,6 +22,11 @@ import (
 	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
 	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
 	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/utils/annotations"
+)
+
+const (
+	enumTargetNameAnnotation = "ebpf.formatter.enum"
```
Comment (flyth) [inline pkg/operators/ebpf/converters.go:29]: nit:
```suggestion

const (
	enumTargetNameAnnotation = "ebpf.converter.enum"
```

Or maybe rename the file.

**Review by flyth — COMMENTED** [review]
Yeah, I think that's a good compromise now, thanks!

File: pkg/utils/annotations/annotations.go (hunk 2 above)
Comment (mauriciovasquezbernal) [reply pkg/utils/annotations/annotations.go:29]: Right, I prefer to keep them separated. I know it's more verbose but also more flexible.

File: pkg/operators/ebpf/converters.go (hunk 3 above)
Comment (mauriciovasquezbernal) [reply pkg/operators/ebpf/converters.go:29]: Renamed the file to have a single formatter concept.

**Review by flyth — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**